### PR TITLE
O3 1302: Implement the Search History component

### PR DESCRIPTION
**Pull request by anjula-sack**
## Purpose

- The purpose of this PR is to fix #[O3-1302](https://issues.openmrs.org/browse/O3-1302)

## Goals

- To give the ability to see the search history

## Approach

- Added the translations
- Fixed the debounce issue
- Added save query and cohort functionality
- Added delete history function
- Added the data download as csv function
- Wrote unit test to test the components

## Screenshots

* Search history getting updated after each search
Kapture 2022-07-03 at 18 49 40

* Download Results
hug

* Save cohort
Kapture 2022-07-03 at 18 54 54

* Save query
nj

* Delete search history item
Kapture 2022-07-03 at 18 57 46

* Delete history
Kapture 2022-07-03 at 19 19 59

### CODE REVIEW

File: src/cohort-builder.tsx [hunk 1]
```diff
@@ -99,13 +88,21 @@ const CohortBuilder: React.FC = () => {
       });
       setPatients(rows);
       addToHistory(queryDescription, rows, searchParams.query);
-      setNotification({
+      showNotification({
+        title: t("success", "Success!"),
         kind: "success",
-        title: `Search is completed with ${rows.length} result(s)`,
+        critical: true,
+        description: `Search is completed with ${rows.length} result(s)`,
```
Comment (jayasanka-sack) [inline src/cohort-builder.tsx:95]: Add translations

File: src/cohort-builder.tsx [hunk 2]
```diff
@@ -99,13 +88,21 @@ const CohortBuilder: React.FC = () => {
...
       setIsLoading(false);
-    } catch (e) {
-      setNotification({ kind: "error", title: e.toString() });
+      setIsHistoryUpdated(true);
+    } catch (error) {
+      showNotification({
+        title: t("error", "Error"),
+        kind: "error",
+        critical: true,
+        description: error?.message,
```
Comment (jayasanka-sack) [inline src/cohort-builder.tsx:104]: would `error.message` come with the selected translation? no, right?

File: src/components/search-concept/search-concept.component.tsx [hunk 3]
```diff
@@ -46,7 +53,17 @@ export const SearchConcept: React.FC<SearchConceptProps> = ({
     }
   };
 
-  const debounceFn = _debounce(onSearch, 500);
+  const debouncedSearch = useRef(
+    _debounce(async (search) => {
+      search != "" && (await onSearch(search));
+    }, 400)
```
Comment (jayasanka-sack) [inline src/components/search-concept/search-concept.component.tsx:59]: `400` milliseconds might not be large enough.

Comment (jayasanka-sack) [reply src/components/search-concept/search-concept.component.tsx:59]: Also, Could you make this more readable?

File: src/components/search-history/search-history.resources.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  );
+
+  return savedCohort;
+}
+
+/**
+ * @returns Query
+ * @param query
+ */
+export async function createQuery(query: Query) {
```
Comment (jayasanka-sack) [inline src/components/search-history/search-history.resources.ts:26]: Missing return type

File: src/components/search-concept/search-concept.component.tsx (hunk 3 above)
Comment (anjula-sack) [reply src/components/search-concept/search-concept.component.tsx:59]: it tested this seems enough, previously debounce setup was different

File: src/cohort-builder.tsx (hunk 2 above)
Comment (anjula-sack) [reply src/cohort-builder.tsx:104]: yeah

File: src/components/search-history/search-history.component.tsx [hunk 5]
```diff
@@ -1,55 +1,225 @@
...
+  SearchHistoryItem,
+} from "../../types/types";
 import EmptyData from "../empty-data/empty-data.component";
+import { createCohort, createQuery } from "./search-history.resources";
 import styles from "./search-history.style.scss";
+import { downloadCSV, getSearchHistory } from "./search-history.utils";
+
+interface SearchHistoryProps {
+  isHistoryUpdated: boolean;
+  setIsHistoryUpdated: Dispatch<SetStateAction<boolean>>;
```
Comment (jayasanka-sack) [inline src/components/search-history/search-history.component.tsx:38]: What is the purpose of these two props?

File: src/components/search-history/search-history.component.tsx [hunk 6]
```diff
@@ -1,55 +1,225 @@
...
+      showNotification({
+        title: t("cohortCreateError", "Error creating the cohort"),
+        kind: "error",
+        critical: true,
+        description: error?.message,
+      });
+    }
+  };
+
+  const handleDeleteSearchItem = async () => {
```
Comment (jayasanka-sack) [inline src/components/search-history/search-history.component.tsx:136]: It's better to take these logics out from the component to make it simple and testable.

**Review by jayasanka-sack — CHANGES_REQUESTED** [review]
Thanks for the PR @anjula-sack ! I added some comments.

File: src/components/search-concept/search-concept.component.tsx [hunk 7]
```diff
@@ -46,7 +53,19 @@ export const SearchConcept: React.FC<SearchConceptProps> = ({
     }
   };
 
-  const debounceFn = _debounce(onSearch, 500);
+  const debouncedSearch = useRef(
+    _debounce(async (searchText: string) => {
+      if (searchText != "") {
```
Comment (jayasanka-sack) [inline src/components/search-concept/search-concept.component.tsx:58]:
```suggestion
      if (searchText) {
```

**Review by jayasanka-sack — APPROVED** [review]
Thanks for getting this done @anjula-sack !

**Pull request merged**